src/Features/CSharp/Portable/AddBraces/CSharpAddBracesDiagnosticAnalyzer.cs

```diff
@@ -86,5 +108,172 @@ public void AnalyzeNode(SyntaxNodeAnalysisContext context)
+            {
+                // The part of the statement preceding the embedded statement does not fit on one line. Examples:
+                //
+                //   for (int i = 0; // <-- The initializer/condition/increment are on separate lines
+                //        i < 10;
+                //        i++)
+                //     SomeMethod();
+                return true;
+            }
+
+            // Check the part of the statement following the embedded statement, but only if it exists
+            if (statement.GetLastToken() != embeddedStatement.GetLastToken())
+            {
+                if (statement is IfStatementSyntax ifStatement && ifStatement.Statement == embeddedStatement)
+                {
+                    // The embedded statement is followed by an 'else' clause, which may span multiple lines without
+                    // triggering a braces requirement, such as this:
+                    //
+                    //   if (true)
+                    //     return;
+                    //   else          // <-- this else clause is two lines, but is not considered a multiline context
+                    //     return;
+                    //
+                    // ---
+                    // INTENTIONAL FALLTHROUGH
+                }
+                else
+                {
+                    var firstTokenAfterEmbeddedStatement = embeddedStatement.GetLastToken().GetNextToken();
+                    if (!FormattingRangeHelper.AreTwoTokensOnSameLine(firstTokenAfterEmbeddedStatement, statement.GetLastToken()))
+                    {
+                        // The part of the statement following the embedded statement does not fit on one line. Examples:
+                        //
+                        //   do
+                        //     SomeMethod();
+                        //   while (x < 0 ||    // <-- This condition is split across multiple lines.
+                        //          x > 10);
+                        return true;
+                    }
+                }
+            }
+
+            return false;
+        }
+
+        /// <summary>
+        /// Determines whether <paramref name="embeddedStatement"/> should use braces under a
+        /// <see cref="PreferBracesPreference.WhenMultiline"/> preference due to the presence of braces on one or more
+        /// sibling statements (the "context").
+        /// </summary>
+        private static bool RequiresBracesToMatchContext(SyntaxNode statement, StatementSyntax embeddedStatement)
+        {
+            if (!statement.IsKind(SyntaxKind.IfStatement, SyntaxKind.ElseClause))
+            {
+                // 'if' statements are the only statements that can have multiple embedded statements which are
+                // considered relative to each other.
+                return false;
+            }
+
+            var topLevelContext = GetTopLevelContext(statement);
```

Comment:
Similarly, this would be "GetFirstIfStatementForContext" or something like that.  The methods, currently, sound vague, but have been spec'ed to only make sense for an if-statement chain. If we needed to add more constructs in the future, we could keep the names vaguer.  But given they're only for if-statements, i would tighten up the names.

Reply:
➡️ Duplicate of https://github.com/dotnet/roslyn/pull/31599#discussion_r239905552